modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaClusterShardingRebalanceListener.scala

```diff
@@ -0,0 +1,164 @@
+
+    def initializeNewKafkaProducerRegions(): ActorState = {
+      val allTopicPartitions = partitionAssignments.partitionAssignments.values.flatten
+
+      allTopicPartitions.foldLeft(this) { case (stateAccum, topicPartition) =>
+        partitionRegionFor(stateAccum, topicPartition.partition).state
+      }
+    }
+
+    private def partitionRegionFor(state: ActorState, partition: Int): StatePlusRegion = {
+      val existingRegionOpt = state.partitionKafkaProducerRegions.get(partition)
+
+      existingRegionOpt.map(region => StatePlusRegion(state, Some(region))).getOrElse(newKafkaProducerRegionForPartition(state, partition))
+    }
+
+    private def newKafkaProducerRegionForPartition(state: ActorState, partition: Int): StatePlusRegion = {
+      state.partitionsToHosts
+        .get(partition)
+        .map { _ =>
+          val kafkaProducerRegion = partitionToKafkaProducerActor(partition)
+          kafkaProducerRegion.start()
+          state.addRegionForPartition(partition, kafkaProducerRegion)
+        }
+        .getOrElse {
+          log.warn(s"Unable to find a partition assignment for partition $partition")
+          StatePlusRegion(state, None)
+        }
+    }
+
+    private def addRegionForPartition(partition: Int, kafkaProducerRegion: KafkaProducerActor): StatePlusRegion = {
+      val newRegionsMap = partitionKafkaProducerRegions + (partition -> kafkaProducerRegion)
+      val newState = this.copy(partitionKafkaProducerRegions = newRegionsMap)
+      StatePlusRegion(newState, Some(kafkaProducerRegion))
+    }
+
+  }
+
+  private val log = LoggerFactory.getLogger(getClass)
+  private val shardAllocationClient = ExternalShardAllocation(context.system).clientFor(groupId)
+  private val address = Cluster(context.system).selfMember.address
+
+  override def preStart(): Unit = {
+    stateTrackingActor ! KafkaConsumerStateTrackingActor.Register(self)
+  }
+
+  override def receive: Receive = uninitialized
+
+  def uninitialized: Receive = { case assignments: PartitionAssignments =>
+    handle(assignments)
+  }
+
+  def initialized(state: ActorState): Receive = { case assignments: PartitionAssignments =>
+    handle(state, assignments)
+  }
+
+  private def handle(partitionAssignments: PartitionAssignments): Unit = {
+    if (partitionAssignments.topicPartitionsToHosts.nonEmpty) {
+      val emptyState = ActorState(PartitionAssignments.empty, Map.empty)
+      handle(emptyState, partitionAssignments)
+    }
+  }
+
+  private def handle(state: ActorState, partitionAssignments: PartitionAssignments): Unit = {
+    val newStateWithPartitionAssignments = state.updatePartitionAssignments(partitionAssignments)
+
+    val updatedTopicPartitions = newStateWithPartitionAssignments.partitionAssignments.partitionAssignments.collect {
+      case (hostPort, topicPartitions) if isHostPortThisNode(hostPort) => topicPartitions
+    }.flatten
+
+    val updates = shardAllocationClient.updateShardLocations(updatedTopicPartitions.map { tp =>
```

Comment:
Should `updateShardLocations` here be updating the locations of all of the shards including those assigned to other nodes in order for the cluster to be aware of which nodes are assigned which shards across the board? Right now it's filtering out and only providing updates on the partitions that are assigned to this node

Reply:
Each node will update its own shard locations and co-locate it with assigned kafka partitions. 
Same thing happens in [akka-kafka-cluster-sharding library](https://github.com/akka/alpakka-kafka/blob/8762635be58b83a4cab9194228343e34b9aa7c60/cluster-sharding/src/main/scala/akka/kafka/cluster/sharding/KafkaClusterSharding.scala#L309) which uses alpakka kafka consumer